### PR TITLE
dbeaver/dbeaver#19383 clickhouse catalog handling

### LINKED ISSUE #19383: Enable catalog support for ClickHouse
**Is your feature request related to a problem? Please describe.**
Both legacy and new ClickHouse JDBC driver use term `schema` as synonymous to database, which is inconsistent with other interfaces(i.e. MySQL) of ClickHouse and most database drivers. Starting from v0.4.2, the driver will support option `databaseTerm`(defaults to schema, can be changed to catalog) to support catalog. In v0.5, the default will be changed from `schema` to `catalog`.

**Describe the solution you'd like**
Change ClickHouse plugin configuration from `<parameter name="omit-catalog" value="true"/>` to the followings:
```xml
<parameter name="omit-catalog" value="true"/>
<parameter name="omit-single-catalog" value="true"/>
```
According to my test, the change works well with v0.4.1. You may verify the change using latest nightly build at [here](https://s01.oss.sonatype.org/content/repositories/snapshots/com/clickhouse/clickhouse-jdbc/0.4.2-SNAPSHOT/).

**Describe alternatives you've considered**
N/A

**Additional context**
Currently the workaround is to change the parameters in driver settings:
image

### PR DESCRIPTION
**Pull request by Destrolaric**
Closes dbeaver/dbeaver#19383 
Removes catalog omitting and adds metadata entity for catalog

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseCatalog.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    void resetStatistics() {
+        hasStatistics = false;
+    }
+
+    @Override
+    public void collectObjectStatistics(DBRProgressMonitor monitor, boolean totalSizeOnly, boolean forceRefresh) throws DBException {
+        if (hasStatistics && !forceRefresh) {
+            return;
+        }
+        try (DBCSession session = DBUtils.openMetaSession(monitor, this, "Read relation statistics")) {
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseCatalog.java]:
```suggestion
        try (DBCSession session = DBUtils.openMetaSession(monitor, this, "Read catalog statistics")) {
```

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseCatalog.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import org.jkiss.dbeaver.model.exec.jdbc.JDBCSession;
+import org.jkiss.dbeaver.model.runtime.DBRProgressMonitor;
+import org.jkiss.dbeaver.model.struct.DBSEntity;
+import org.jkiss.dbeaver.model.struct.DBSObject;
+
+import java.sql.SQLException;
+import java.util.List;
+
+public class ClickhouseCatalog extends GenericCatalog implements DBPObjectStatisticsCollector, DBPSystemObject
+{
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseCatalog.java]:
```suggestion
public class ClickhouseCatalog extends GenericCatalog implements DBPObjectStatisticsCollector, DBPSystemObject {
```

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseCatalog.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    private boolean hasStatistics = true;
+
+
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseCatalog.java:43]:
```suggestion

```

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseCatalog.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                    "select table," +
+                        "sum(bytes) as table_size, " +
+                        "sum(rows) as table_rows, " +
+                        "max(modification_time) as latest_modification," +
+                        "min(min_date) AS min_date," +
+                        "max(max_date) AS max_date " +
+                        "FROM system.parts\n" +
+                        "WHERE database=? AND active\n" +
+                        "GROUP BY table"))
+            {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseCatalog.java]:
```suggestion
                        "GROUP BY table"
            )) {
```

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseCatalog.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    @Override
+    public void collectObjectStatistics(DBRProgressMonitor monitor, boolean totalSizeOnly, boolean forceRefresh) throws DBException {
+        if (hasStatistics && !forceRefresh) {
+            return;
+        }
+        try (DBCSession session = DBUtils.openMetaSession(monitor, this, "Read catalog statistics")) {
+            try (
+                JDBCPreparedStatement dbStat = ((JDBCSession)session).prepareStatement(
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseCatalog.java]:
```suggestion
                JDBCPreparedStatement dbStat = ((JDBCSession) session).prepareStatement(
```

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseCatalog.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        return hasStatistics;
+    }
+
+    @Override
+    public synchronized DBSObject refreshObject(@NotNull DBRProgressMonitor monitor) throws DBException {
+        resetStatistics();
+        return super.refreshObject(monitor);
+    }
+
+    void resetStatistics() {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseCatalog.java:69]: Can be `private`

**Review by ShadelessFox — APPROVED** [review]

**Review by E1izabeth — APPROVED** [review]

**Comment by LonwoLonwo** [comment]
Does it work for the legacy Clickhouse driver? Can you read table data in both drivers? Can you read tables data with schemas and catalogs?

**Comment by Destrolaric** [comment]
@LonwoLonwo Legacy driver will use only schemas. It doesn't have a feature to show them as a catalog. This is one of the recent features of the driver.

**Comment by Matvey16** [comment]
Verified

**Pull request merged**